pkg/storage/snapshot/restore.go

```diff
@@ -596,6 +606,117 @@ func (t *vmRestoreTarget) reconcileDataVolumes() (bool, error) {
+	vmSnapshotKey := cacheKeyFunc(namespace, name)
+	obj, exists, err := t.controller.VMSnapshotInformer.GetStore().GetByKey(vmSnapshotKey)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("Unable to find VirtualMachineSnapshot %s", vmSnapshotKey)
+	}
+	return obj.(*snapshotv1.VirtualMachineSnapshot), nil
+}
+
+func (t *vmRestoreTarget) restoreInstancetypeControllerRevision(VMSnapshotRevisionName, vmSnapshotName string, vm *kubevirtv1.VirtualMachine) (*appsv1.ControllerRevision, error) {
+	// Switch the snapshot and vm names for the restored CR name
+	restoredCRName := vm.Name + strings.Replace(VMSnapshotRevisionName, vmSnapshotName, "", 1)
+
+	// If the target VirtualMachine already exists it's likely that the original ControllerRevision is already present.
+	// Check that here by attempting to lookup the CR using the generated restoredCRName.
+	// Ignore any NotFound errors raised allowing the CR to be restored below.
+	if t.doesTargetVMExist() {
+		existingCR, err := t.getControllerRevision(vm.Namespace, restoredCRName)
+		if err != nil && !errors.IsNotFound(err) {
+			return nil, err
+		}
+		if existingCR != nil {
+			// TODO - Check the contents of the existing CR here against that of the snapshot CR
+			return existingCR, nil
+		}
+	}
+
+	snapshotCR, err := t.getControllerRevision(vm.Namespace, VMSnapshotRevisionName)
+	if err != nil {
+		return nil, err
+	}
+
+	restoredCR := snapshotCR.DeepCopy()
+	restoredCR.ObjectMeta.Reset()
+	restoredCR.Name = restoredCRName
+
+	restoredCR, err = t.controller.Client.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), restoredCR, metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	return restoredCR, nil
+}
+
+func (t *vmRestoreTarget) restoreInstancetypeControllerRevisions(vm *kubevirtv1.VirtualMachine) error {
+	if vm.Spec.Instancetype != nil && len(vm.Spec.Instancetype.RevisionName) > 0 {
+		restoredCR, err := t.restoreInstancetypeControllerRevision(vm.Spec.Instancetype.RevisionName, t.vmRestore.Spec.VirtualMachineSnapshotName, vm)
+		if err != nil {
+			return err
+		}
+		vm.Spec.Instancetype.RevisionName = restoredCR.Name
+	}
+
+	if vm.Spec.Preference != nil && len(vm.Spec.Preference.RevisionName) > 0 {
+		restoredCR, err := t.restoreInstancetypeControllerRevision(vm.Spec.Preference.RevisionName, t.vmRestore.Spec.VirtualMachineSnapshotName, vm)
+		if err != nil {
+			return err
+		}
+		vm.Spec.Preference.RevisionName = restoredCR.Name
+	}
+
+	return nil
+}
+
+func (t *vmRestoreTarget) claimInstancetypeControllerRevisionOwnership(revisionName string, vm *kubevirtv1.VirtualMachine) error {
+	cr, err := t.getControllerRevision(vm.Namespace, revisionName)
+	if err != nil {
+		return err
+	}
+
+	if !metav1.IsControlledBy(cr, vm) {
+		cr.OwnerReferences = []metav1.OwnerReference{*metav1.NewControllerRef(vm, kubevirtv1.VirtualMachineGroupVersionKind)}
```

Comment:
Shouldn't we instead append the VM to `cr.OwnerReferences`? What if the CR is also owned by a snapshot object?

Reply:
That isn't a possibility here, the CR referenced by the restored VM we are looking at here can only be owned by the restored VM.